Compilation with NVCC without CUDA

I have a bit of an unusual setup where I compile my code multiple times for multiple kokkos targets to create multiple shared libraries. This worked just fine but fails now with the recent update to 3.6. The root cause being the definition of `KOKKOS_IF_ON_DEVICE`.

https://github.com/kokkos/kokkos/blob/b52f8c835f4df003954dad66d9761094f8baa66c/core/src/Kokkos_Macros.hpp#L567-L578

When I'm compiling my code with an OpenMP version of Kokkos using NVCC `__CUDA_ARCH__` will be set resulting in a non-empty definition of `KOKKOS_IF_ON_DEVICE`. Thus, the initial `#if` should be at least something like

```
#if (defined(__CUDA_ARCH__) && defined(KOKKOS_ENABLE_CUDA)) || \
    defined(__HIP_DEVICE_COMPILE__) || \
    defined(__SYCL_DEVICE_ONLY__)
```

maybe similar guards are needed for HIP and SYCL, but since I'm not familiar with them, I don't want to add any wrong code to my suggestion.

```
+ cmake -DCMAKE_BUILD_TYPE=Debug -DCMAKE_CXX_COMPILER_LAUNCHER=ccache -DCMAKE_CXX_COMPILER=/var/jenkins/workspace/Kokkos/bin/nvcc_wrapper -DCMAKE_CXX_FLAGS=-Werror -DCMAKE_CXX_STANDARD=14 -DKokkos_ARCH_NATIVE=ON -DKokkos_ENABLE_COMPILER_WARNINGS=ON -DKokkos_ENABLE_DEBUG=ON -DKokkos_ENABLE_DEBUG_BOUNDS_CHECK=ON -DKokkos_ENABLE_DEPRECATED_CODE_3=OFF -DKokkos_ENABLE_TESTS=ON -DKokkos_ENABLE_LIBDL=OFF ..
-- Setting default Kokkos CXX standard to 14
-- The CXX compiler identification is GNU 9.4.0
-- Check for working CXX compiler: /var/jenkins/workspace/Kokkos/bin/nvcc_wrapper
-- Check for working CXX compiler: /var/jenkins/workspace/Kokkos/bin/nvcc_wrapper -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Could NOT find GTest (missing: GTEST_LIBRARY GTEST_INCLUDE_DIR GTEST_MAIN_LIBRARY) 
-- The project name is: Kokkos
-- SERIAL backend is being turned on to ensure there is at least one Host space. To change this, you must enable another host execution space and configure with -DKokkos_ENABLE_SERIAL=OFF or change CMakeCache.txt
-- Using -std=gnu++14 for C++14 extensions as feature
-- Built-in Execution Spaces:
--     Device Parallel: NoTypeDefined
--     Host Parallel: NoTypeDefined
--       Host Serial: SERIAL
-- 
-- Architectures:
--  NATIVE
-- Using internal desul_atomics copy
-- Kokkos Devices: SERIAL, Kokkos Backends: SERIAL
-- Configuring done
-- Generating done
-- Build files have been written to: /var/jenkins/workspace/Kokkos/build
+ make -j8
Scanning dependencies of target kokkos_gtest
[  0%] Building CXX object core/unit_test/CMakeFiles/kokkos_gtest.dir/__/__/tpls/gtest/gtest/gtest-all.cc.o
Scanning dependencies of target kokkoscore
[  0%] Building CXX object core/src/CMakeFiles/kokkoscore.dir/impl/Kokkos_Command_Line_Parsing.cpp.o
[  1%] Building CXX object core/src/CMakeFiles/kokkoscore.dir/impl/Kokkos_CPUDiscovery.cpp.o
[  1%] Building CXX object core/src/CMakeFiles/kokkoscore.dir/impl/Kokkos_Core.cpp.o
[  1%] Building CXX object core/src/CMakeFiles/kokkoscore.dir/impl/Kokkos_ExecPolicy.cpp.o
[  1%] Building CXX object core/src/CMakeFiles/kokkoscore.dir/impl/Kokkos_Error.cpp.o
[  1%] Building CXX object core/src/CMakeFiles/kokkoscore.dir/impl/Kokkos_HostBarrier.cpp.o
[  2%] Building CXX object core/src/CMakeFiles/kokkoscore.dir/impl/Kokkos_HostSpace.cpp.o
nvcc_wrapper has been given GNU extension standard flag -std=gnu++14 - reverting flag to -std=c++14
nvcc warning : The 'compute_35', 'compute_37', 'sm_35', and 'sm_37' architectures are deprecated, and may be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning).
nvcc_wrapper has been given GNU extension standard flag -std=gnu++14 - reverting flag to -std=c++14
nvcc warning : The 'compute_35', 'compute_37', 'sm_35', and 'sm_37' architectures are deprecated, and may be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning).
/var/jenkins/workspace/Kokkos/core/src/impl/Kokkos_Error.hpp(232): error: namespace "Kokkos::Impl" has no member "device_abort"

/var/jenkins/workspace/Kokkos/core/src/impl/Kokkos_Error.hpp(233): warning #1305-D: function declared with "noreturn" does return

1 error detected in the compilation of "/var/jenkins/workspace/Kokkos/core/src/impl/Kokkos_Error.cpp".
make[2]: *** [core/src/CMakeFiles/kokkoscore.dir/build.make:105: core/src/CMakeFiles/kokkoscore.dir/impl/Kokkos_Error.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
/var/jenkins/workspace/Kokkos/core/src/impl/Kokkos_Error.hpp(232): error: namespace "Kokkos::Impl" has no member "device_abort"

/var/jenkins/workspace/Kokkos/core/src/impl/Kokkos_Error.hpp(233): warning #1305-D: function declared with "noreturn" does return

nvcc_wrapper has been given GNU extension standard flag -std=gnu++14 - reverting flag to -std=c++14
nvcc warning : The 'compute_35', 'compute_37', 'sm_35', and 'sm_37' architectures are deprecated, and may be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning).
nvcc_wrapper has been given GNU extension standard flag -std=gnu++14 - reverting flag to -std=c++14
nvcc warning : The 'compute_35', 'compute_37', 'sm_35', and 'sm_37' architectures are deprecated, and may be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning).
1 error detected in the compilation of "/var/jenkins/workspace/Kokkos/core/src/impl/Kokkos_HostSpace.cpp".
```
Is that the error you are seeing?

--- a/core/src/Kokkos_Macros.hpp
+++ b/core/src/Kokkos_Macros.hpp
@@ -564,8 +564,9 @@ static constexpr bool kokkos_omp_on_host() { return false; }
 #endif
 
 #if !defined(KOKKOS_IF_ON_HOST) && !defined(KOKKOS_IF_ON_DEVICE)
-#if defined(__CUDA_ARCH__) || defined(__HIP_DEVICE_COMPILE__) || \
-    defined(__SYCL_DEVICE_ONLY__)
+#if (defined(KOKKOS_ENABLE_CUDA) && defined(__CUDA_ARCH__)) ||         \
+    (defined(KOKKOS_ENABLE_HIP) && defined(__HIP_DEVICE_COMPILE__)) || \
+    (defined(KOKKOS_ENABLE_SYCL) && defined(__SYCL_DEVICE_ONLY__))
 #define KOKKOS_IF_ON_DEVICE(CODE) \
   { KOKKOS_IMPL_STRIP_PARENS(CODE) }
 #define KOKKOS_IF_ON_HOST(CODE) \

--- a/core/unit_test/TestNumericTraits.hpp
+++ b/core/unit_test/TestNumericTraits.hpp
@@ -213,8 +213,8 @@ struct TestNumericTraits {
   }
 };
 
-#if defined(KOKKOS_COMPILER_NVCC) || defined(KOKKOS_ENABLE_SYCL) || \
-    defined(KOKKOS_ENABLE_OPENMPTARGET)
+#if (defined(KOKKOS_COMPILER_NVCC) && defined(KOKKOS_ENABLE_CUDA)) || \
+    defined(KOKKOS_ENABLE_SYCL) || defined(KOKKOS_ENABLE_OPENMPTARGET)
 template <class Tag>
 struct TestNumericTraits<
 #if defined(KOKKOS_ENABLE_CUDA)
@@ -600,8 +600,9 @@ CHECK_SAME_AS_NUMERIC_LIMITS_MEMBER_CONSTANT(long double, max_exponent10);
 // Workaround compiler issue error: expression must have a constant value
 // See kokkos/kokkos#4574
 // There is the same bug with CUDA 11.6
-// FIXME_NVHPC FIXME_CUDA
-#if !defined(KOKKOS_COMPILER_NVHPC) && (CUDA_VERSION < 11060)
+// FIXME_NVHPC FIXME_CUDA FIXME_NVCC
+#if !defined(KOKKOS_COMPILER_NVHPC) && (CUDA_VERSION < 11060) && \
+    !(defined(KOKKOS_COMPILER_NVCC) && !defined(KOKKOS_ENABLE_CUDA))
 CHECK_NAN_SAME_AS_NUMERIC_LIMITS_MEMBER_FUNCTION(float, quiet_NaN);
 CHECK_NAN_SAME_AS_NUMERIC_LIMITS_MEMBER_FUNCTION(double, quiet_NaN);
 CHECK_NAN_SAME_AS_NUMERIC_LIMITS_MEMBER_FUNCTION(long double, quiet_NaN);

